### PR TITLE
Reduce logfile output in HEMCO, and combine Verbose and Warnings into a single switch

### PR DESCRIPTION
**Pull request by yantosca**
This is the PR corresponding to issue #180 and to https://github.com/geoschem/geos-chem/pull/1549.

In this PR we seek to:
1. Convert most HEMCO logfile output to debug-only printout
2. Consolidate Verbose and Warnings into a single logical switch in `HEMCO_Config.rc` (e.g. `Verbose: true`) so that warnings will also be toggled by Verbose.  There isn't much need for separate verbose & warnings levels any more. 

Tagging @lizziel @msulprizio

### CODE REVIEW

**Comment by yantosca** [comment]
Also tagging @christophkeller

**Comment by yantosca** [comment]
This PR is now ready for review.

**Comment by yantosca** [comment]
Integration tests pending (with GEOS-Chem)

HEMCO standalone simulation log output (24hr run)
- with `Verbose: true:`  [HEMCO_sa.log.verbose.txt](https://github.com/geoschem/HEMCO/files/10209876/HEMCO_sa.log.verbose.txt)
- with `Verbose: false`:  [HEMCO_sa.log.noverbose.txt](https://github.com/geoschem/HEMCO/files/10209882/HEMCO_sa.log.noverbose.txt)

**Review by msulprizio — CHANGES_REQUESTED** [review]

**Comment by yantosca** [comment]
@jimmielin: I have updated the program logic (in 5041a2000d5d35eb4670fa49303b34d28309edd3) so that HEMCO verbose will be toggled either with Verbose true/false, or with Verbose and Warnings having integer values.  See the section here:

```F90

! !LOCAL VARIABLES:
!
    ! Scalars
    LOGICAL               :: doVerbose,   found
    LOGICAL               :: foundVerb,   foundWarn
    LOGICAL               :: verboseBool
    INTEGER               :: I,           N,         POS
    INTEGER               :: verb,        warn

    ! Strings
    CHARACTER(LEN=255)    :: line
    CHARACTER(LEN=255)    :: loc
    CHARACTER(LEN=255)    :: LogFile
    CHARACTER(LEN=255)    :: DiagnPrefix
    CHARACTER(LEN=255)    :: MetField
    CHARACTER(LEN=255)    :: GridRes
    CHARACTER(LEN=512)    :: msg

    !======================================================================
    ! ReadSettings begins here
    !======================================================================

    ! Enter
    Loc = 'ReadSettings (hco_config_mod.F90)'

    ... etc ...
    !-----------------------------------------------------------------------
    ! Read settings and add them as options to core extensions
    !-----------------------------------------------------------------------

    ! Do until exit
    DO

       ! Read line
       CALL HCO_ReadLine ( IU_HCO, LINE, EOF, RC )
       IF ( RC /= HCO_SUCCESS ) THEN
          msg = 'Error in HEMCO_Config.rc @ line: ' // TRIM( Line )
          CALL HCO_Error( msg, RC, thisLoc=loc )
          RETURN
       ENDIF

       ! Return if EOF
       IF ( EOF ) EXIT

       ! Test if the Verbose flag is logical.  We need to know this ahead
       ! of time in order to avoid input errors (i.e. reading characters
       ! when integers are expected) in the code that follows below.
       !  -- Bob Yantosca (14 Dec 2022)
       IF ( INDEX( line, 'Verbose' ) > 0 ) THEN
          verboseBool = (                                                    &
             ( INDEX( line, 't' ) > 0 ) .or. ( INDEX( line, 'f' ) > 0 ) .or. &
             ( INDEX( line, 'T' ) > 0 ) .or. ( INDEX( line, 'F' ) > 0 )     )
       ENDIF

       ... etc...

       IF ( TRIM(LINE) == '' ) CYCLE

       ! Add this option to HEMCO core
       CALL AddExtOpt ( HcoConfig, TRIM(LINE), &
                        CoreNr, RC, IgnoreIfExist=.TRUE. )
       IF ( RC /= HCO_SUCCESS ) THEN
          msg = 'Error in HEMCO_Config.rc @ line: ' // TRIM( Line )
          CALL HCO_Error( msg, RC, thisLoc=loc )
          RETURN
       ENDIF

    ENDDO

    ... etc ...

    !-----------------------------------------------------------------------
    ! Initialize error object if needed.
    ! Extract values to initialize error module and set some further
    ! HEMCO variables. Only the first time the settings are read (settings
    ! can be read multiple times if nested HEMCO configuration files are
    ! used)
    !
    ! NOTE: In HEMCO 3.7.0, the Verbose and Warnings integers in the
    ! HEMCO_Config.rc file have been replaced with "Verbose: true".
    ! Update the logic to make the test for Verbose backwards compatible
    ! with HEMCO_Config.files prior to HEMCO 3.7.0.
    !   -- Bob Yantosca (14 Dec 2022)
    !-----------------------------------------------------------------------
    IF ( .NOT. ASSOCIATED(HcoConfig%Err) ) THEN

       ! Initialize
       doVerbose = .FALSE.
       verb      = 0
       warn      = 0

       ! Check if Verbose is a logical entry in HEMCO_Config.rc
       IF ( verboseBool ) THEN

          !-----------------------------------------------------------------
          ! "Verbose: true" or "Verbose: false" was found
          !-----------------------------------------------------------------

          ! First look for Verbose (logical).  This is now the default
          ! inthe HEMCO_Config.rc file for HEMCO 3.7.0 and later.
          CALL GetExtOpt( HcoConfig,            CoreNr,       'Verbose',     &
                          OptValBool=doVerbose, found=found,  RC=RC         )
          IF ( RC /= HCO_SUCCESS ) THEN
             msg = 'Error looking for "Verbose" (logical) in HEMCO_Config.rc!'
             CALL HCO_Error( msg, RC, thisLoc=loc )
             RETURN
          ENDIF

          ! Print status message
          IF ( doVerbose ) THEN
             msg = NEW_LINE( 'A' ) // 'HEMCO verbose output is ON'
          ELSE
             msg = NEW_LINE( 'A' ) // 'HEMCO verbose output is OFF'
          ENDIF
          CALL HCO_Msg( msg, verb=.TRUE. )

       ELSE

          !-----------------------------------------------------------------
          ! "Verbose: true" or "Verbose: false" was not found
          !-----------------------------------------------------------------

          ! Check for Verbose (integer)
          CALL GetExtOpt( HcoConfig,      CoreNr,          'Verbose',        &
                          OptValInt=verb, found=foundVerb,  RC=RC           )
          IF ( RC /= HCO_SUCCESS ) THEN
             msg = 'Error looking for "Verbose" (integer) in HEMCO_Config.rc!'
             CALL HCO_Error( msg, RC, thisLoc=loc )
             RETURN
          ENDIF

          IF ( foundVerb .and. verb > 0 ) THEN

             ! Toggle HEMCO verbose output on for nonzero integer values
             doVerbose = .TRUE.
             msg = NEW_LINE( 'A' )                                        // &
                  'HEMCO verbose output is ON.'                           // &
                   NEW_LINE( 'A' )                                        // &
                  'Numbered Verbose and Warning options are deprecated.'  // &
                   NEW_LINE( 'A' )                                        // &
                  'Please use "Verbose: true:" or "Verbose: false" for '  // &
                  'controlling verbose output.'
             CALL HCO_Msg( msg, verb=.TRUE. )

          ELSE

             ! Verbose (logical) and Verbose (integer) were not found,
             ! now look for Warnings (integer)
             CALL GetExtOpt( HcoConfig,      CoreNr,         'Warnings',     &
                             OptValInt=warn, found=foundWarn, RC=RC         )
             IF ( RC /= HCO_SUCCESS ) THEN
                msg = &
                 'Error looking for "Warnings" (integer) in HEMCO_Config.rc!'
                CALL HCO_Error( msg, RC, thisLoc=loc )
                RETURN
             ENDIF
             IF ( foundWarn .and. warn > 0 ) THEN

                ! Toggle HEMCO verbose output on
                ! (Verbose = 0; Warnings > 0)
                doVerbose = .TRUE.
                msg =                                                        &
                   NEW_LINE( 'A' )                                        // &
                  'HEMCO verbose output is ON.'                           // &
                   NEW_LINE( 'A' )                                        // &
                  'Numbered Verbose and Warning options are deprecated.'  // &
                   NEW_LINE( 'A' )                                        // &
                  'Please use "Verbose: true:" or "Verbose: false" for '  // &
                  'controlling verbose output.'
                CALL HCO_Msg( msg, verb=.TRUE. )

             ELSE

                ! Toggle HEMCO verbose off
                ! (Verbose: false, Verbose = 0 and Warnings = 0)
                doVerbose = .FALSE.
                msg =                                                        &
                   NEW_LINE( 'A' )                                        // &
                  'HEMCO verbose output is OFF.'                          // &
                   NEW_LINE( 'A' )                                        // &
                  'Numbered Verbose and Warning options are deprecated.'  // &
                   NEW_LINE( 'A' )                                        // &
                  'Please use "Verbose: true:" or "Verbose: false" for '  // &
                  'controlling verbose output.'
                CALL HCO_Msg( msg, verb=.TRUE. )
             ENDIF
          ENDIF
       ENDIF

   ... etc ...
```
TL;DR: 
1. When reading each line from `HEMCO_Config.rc`, it checks to see if `Verbose` is logical (`t`, `T`, `f`, `F` in the same line with Verbose) or not.
2. If `Verbose` is logical, then we get the value of `Verbose` into a logical variable `doVerbose`.  We then print a message about whether or not verbose output is on.
3. If `Verbose` is not logical, then try to read the integer values of `Verbose` and, if necessary `Warnings`.  If either of these are nonzero, verbose output will be turned on.  We print a message about the status of verbose output (on/off) and then also a note about the integer values being deprecated.

I ran a bunch of tests with the various settings of `Verbose` and `Warnings`, and show the message that is printed out.

```console
-----------------------------------------------------------------------------------
Verbose:                     0
Warnings:                    0

HEMCO verbose output is OFF.
Numbered Verbose and Warning options are deprecated.
Please use "Verbose: true:" or "Verbose: false" for controlling verbose output.

-----------------------------------------------------------------------------------
Verbose:                     1
Warnings:                    0

HEMCO verbose output is ON.
Numbered Verbose and Warning options are deprecated.
Please use "Verbose: true:" or "Verbose: false" for controlling verbose output.

-----------------------------------------------------------------------------------
Verbose:                     2
Warnings:                    0

HEMCO verbose output is ON.
Numbered Verbose and Warning options are deprecated.
Please use "Verbose: true:" or "Verbose: false" for controlling verbose output.

-----------------------------------------------------------------------------------
Verbose:                     3
Warnings:                    0

HEMCO verbose output is ON.
Numbered Verbose and Warning options are deprecated.
Please use "Verbose: true:" or "Verbose: false" for controlling verbose output.

-----------------------------------------------------------------------------------
Verbose:                     0
Warnings:                    1

HEMCO verbose output is ON.
Numbered Verbose and Warning options are deprecated.
Please use "Verbose: true:" or "Verbose: false" for controlling verbose output.

-----------------------------------------------------------------------------------
Verbose:                     0
Warnings:                    2

HEMCO verbose output is ON.
Numbered Verbose and Warning options are deprecated.
Please use "Verbose: true:" or "Verbose: false" for controlling verbose output.


-----------------------------------------------------------------------------------
Verbose:                     true

HEMCO verbose output is ON

-----------------------------------------------------------------------------------
Verbose:                     false

HEMCO verbose output is OFF

-----------------------------------------------------------------------------------
no verbose in HEMCO_Config.rc

HEMCO verbose output is OFF

-----------------------------------------------------------------------------------
```
So I think this should be OK.

NOTE: I had to pre-determine whether `Verbose` is logical or not in order to avoid runtime read errors (i.e. reading integer values and getting characters, etc.)

File: src/Core/hco_config_mod.F90 [hunk 1]
```diff
@@ -1991,9 +1991,10 @@ SUBROUTINE ReadSettings( HcoConfig, IU_HCO, EOF, RC )
 ! !LOCAL VARIABLES:
 !
     ! Scalars
+    LOGICAL               :: doVerbose
     LOGICAL               :: FOUND
     INTEGER               :: I, N, POS
-    INTEGER               :: verb
+   !INTEGER               :: verb
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_config_mod.F90]: We now actually use this variable in the new program logic (see commit 5041a2000d5d35eb4670fa49303b34d28309edd3)

File: src/Core/hco_config_mod.F90 [hunk 2]
```diff
@@ -2105,16 +2106,31 @@ SUBROUTINE ReadSettings( HcoConfig, IU_HCO, EOF, RC )
...
+       !IF ( RC /= HCO_SUCCESS ) THEN
+       !   msg = 'Error looking for "Verbose" HEMCO_Config.rc!'
+       !   CALL HCO_Error( msg, RC, thisLoc=loc )
+       !   RETURN
+       !ENDIF
+       !IF ( .NOT. FOUND ) THEN
+       !   verb = 3
+       !   WRITE(*,*) 'Setting `Verbose` not found in HEMCO logfile - use 3'
+       !ENDIF
+       !--------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_config_mod.F90]: This has been removed in the updating of program logic (see commit 5041a2000d5d35eb4670fa49303b34d28309edd3).

File: src/Core/hco_config_mod.F90 [hunk 3]
```diff
@@ -2131,18 +2147,22 @@ SUBROUTINE ReadSettings( HcoConfig, IU_HCO, EOF, RC )
...
+       !IF ( RC /= HCO_SUCCESS ) THEN
+       !   msg = 'Error looking for "Warnings" in HEMCO_Config.rc!'
+       !   CALL HCO_Error( msg, RC, thisLoc=loc )
+       !   RETURN
+       !ENDIF
+       !IF ( .NOT. FOUND ) THEN
+       !   warn = 3
+       !   WRITE(*,*) 'Setting `Warnings` not found in HEMCO logfile - use 3'
+       !ENDIF
+       !---------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_config_mod.F90]: This has been removed in the updating of program logic (see commit https://github.com/geoschem/HEMCO/commit/5041a2000d5d35eb4670fa49303b34d28309edd3).

File: src/Core/hcoio_read_std_mod.F90 [hunk 4]
```diff
@@ -1094,14 +1094,15 @@ SUBROUTINE HCOIO_Read( HcoState, Lct, RC )
        ENDIF
 
        ! Prompt a warning if thisUnit is not recognized as unitless.
+       ! TODO: Merge warnings w/ verbose, and toggle with a simple on/off
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hcoio_read_std_mod.F90]: Now removed in 098f91ea3979eb4a39315803ea66756fda92a767

File: src/Core/hco_error_mod.F90 [hunk 5]
```diff
@@ -1013,24 +1000,25 @@ SUBROUTINE HCO_LogFile_Open( Err, RC )
 
     ! Write header on first call
     IF ( Err%FirstOpen ) THEN
-       IF ( Err%LUN < 0 ) THEN
-          LUN = 6                ! Log gets written to stdout
-       ELSE
-          LUN = Err%LUN          ! Log gets written to file
-       ENDIF
+       LUN = Err%Lun                ! Log gets written to file
+       IF ( Err%LUN < 0 ) LUN = 6   ! ,,, or to stdout if file isn't open
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_error_mod.F90]: Now removed in 098f91ea3979eb4a39315803ea66756fda92a767

File: src/Core/hco_error_mod.F90 [hunk 6]
```diff
@@ -888,11 +872,14 @@ FUNCTION HCO_IsVerb ( Err, VerbNr ) RESULT ( IsVerb )
...
-    ELSE
-       IsVerb = ( Err%Verbose >= VerbNr )
-    ENDIF
+    ! Initialize
+    isVerb = .FALSE.
+
+    ! Return if the Err object is null
+    IF ( .not. ASSOCIATED( Err ) ) RETURN
+
+    ! Check if "Verbose: 3" was set in the HEMCO_Config.rc file
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_error_mod.F90:829]: Now removed in 098f91ea3979eb4a39315803ea66756fda92a767

File: src/Core/hco_error_mod.F90 [hunk 7]
```diff
@@ -858,28 +844,26 @@ END FUNCTION HCO_VERBOSE_INQ
...
 ! !OUTPUT PARAMETERS:
 !
-    LOGICAL                   :: IsVerb
+    LOGICAL                :: isVerb
 !
-! !REVISION HISTORY:
-!  15 Mar 2015 - C. Keller - Initialization
-!  See https://github.com/geoschem/hemco for complete history
+! !REMARKS:
+!  TODO: Convert VERBOSE to a simple logical on/off switch.
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_error_mod.F90]: Now removed in 098f91ea3979eb4a39315803ea66756fda92a767

File: src/Core/hco_error_mod.F90 [hunk 8]
```diff
@@ -858,28 +844,26 @@ END FUNCTION HCO_VERBOSE_INQ
 !------------------------------------------------------------------------------
 !BOP
 !
-! !IROUTINE: HCO_IsVerb
+! !IROUTINE: HCO_IsVerb_NoVerb
 !
-! !DESCRIPTION: Function HCO\_IsVerb returns true if the HEMCO verbose number
-!  is equal to or larger than the passed number.
+! !DESCRIPTION: Returns true if the HEMCO verbose number is set to 3 or larger.
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_error_mod.F90]: Now updated in 098f91ea3979eb4a39315803ea66756fda92a767

File: src/Core/hco_error_mod.F90 [hunk 9]
```diff
@@ -740,20 +725,21 @@ SUBROUTINE HCO_ERROR_SET( am_I_Root, Err, LogFile, &
     ! Set verbose to -1 if this is not the root CPU. This will disable any
     ! log-file messages
     IF ( .NOT. am_I_Root ) THEN
-       Verbose      = -1
-       WarningLevel =  0
+       !Verbose      = -1
+       !WarningLevel =  0
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_error_mod.F90]: Now removed in 098f91ea3979eb4a39315803ea66756fda92a767

File: src/Core/hco_config_mod.F90 [hunk 10]
```diff
@@ -2158,9 +2178,12 @@ SUBROUTINE ReadSettings( HcoConfig, IU_HCO, EOF, RC )
        IF ( TRIM(LogFile) == HCO_GetOpt(HcoConfig%ExtList,'Wildcard') )      &
             LogFile = '*'
 
+       !! We should now have everything to define the HEMCO error settings
+       !CALL HCO_ERROR_SET( HcoConfig%amIRoot, HcoConfig%Err, LogFile, &
+       !                    verb, warn, RC )
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_config_mod.F90]: Now removed in 098f91ea3979eb4a39315803ea66756fda92a767

**Review by msulprizio — APPROVED** [review]

**Comment by yantosca** [comment]
I have removed much of the prior logic in commits 824b07f0489bbf322ac77b650127c2d2071d62b2 and 78238dfc9e180d2a7f10e69763d00c284b9d1741.

Also, in commit https://github.com/geoschem/geos-chem/pull/1549/commits/0df54f6dc28b2abefa9d038c7f4935435ff2b547 I have added the `VerboseOnCores` setting to allow users to specify if they would like to send verbose output to the root core or to all cores (analogous to GEOS-Chem).

Integration tests running

**Comment by yantosca** [comment]
I have merged the logfile cleanup updates to GEOS-Chem 14.1.0 and HEMCO 3.6.0.  New integration tests are running.
